docs/source/core_concepts/index.rst

```diff
@@ -0,0 +1,522 @@
+for each output arc, which input arc it corresponds to.
+
+For example, in composition an output arc would usually come from a pair
+of arcs, one in each input FSA.
+
+.. HINT::
+
+  arc map and autograd are implementation details and are not visible to Python
+  API users.
+
+In the following we give two examples about autograd with the following FSA
+in the context of computing total scores with tropical semiring and log semiring.
+
+
+.. _autograd example:
+.. figure:: images/autograd.svg
+    :alt: An FSA for demonstrating autograd.
+    :align: center
+    :figwidth: 600px
+
+    An example FSA for demonstrating autograd in k2.
+
+Arc scores ``a``, ``b``, ``c``, and ``d`` are some numbers not known yet.
+They can come from the output of some neural network and their value depends
+on the internal parameters of the neural network which are updated
+by some gradient descent based algorithms.
+
+Example 1: Autograd in tropical semiring
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The following code shows how to compute the best score of the shortest path for
+the FSA given in :numref:`autograd example`:
+
+.. code-block:: python
+
+  import k2
+
+  nnet_output = torch.tensor([0.1, 1, 0.2, 0.5], dtype=torch.float32)
+  # assume nnet_output is the output of some neural network
+  nnet_output.requires_grad_(True)
+  s = '''
+  0 1 10 0
+  0 2 20 0
+  1 3 -1 0
+  2 3 -1 0
+  3
+  '''
+  fsa = k2.Fsa.from_str(s)
+  fsa.scores = nnet_output
+  fsa.draw('autograd_tropical.svg')
+  fsa_vec = k2.create_fsa_vec([fsa])
+  total_scores = k2.get_tot_scores(fsa_vec, log_semiring=False, use_double_scores=False)
+
+  total_scores.backward()
+  print(nnet_output.grad)
+  # It prints: tensor([0., 1., 0., 1.])
+
+.. figure:: images/autograd_tropical.svg
+    :alt: An example FSA for autograd with tropical scores
+    :align: center
+    :figwidth: 600px
+
+    Output of the above code: autograd_tropical.svg
+
+**Explanation**:
+    - We assume that ``nnet_output = torch.tensor([a, b, c, d]) = torch.tensor([0.1, 1, 0.2, 0.5])``
+      and we set ``nnet_output.requires_grad_(True)`` to simulate that it comes from the output of
+      some neural network.
+
+    - Arc 0: state 0 -> state 1, with score 0.1
+    - Arc 1: state 0 -> state 2, with score 1
+    - Arc 2: state 1 -> state 3, with score 0.2
+    - Arc 3: state 2 -> state 3, witch score 0.5
+    - Score of path 0: arc 0 -> arc 2 is 0.1 + 0.2 = 0.3
+    - Score of path 1: arc 1 -> arc 3 is 1 + 0.5 = 1.5
+    - The best path consists of arc 1 and arc 3.
+    - The best score is ``s = b + d = 1.5``
+
+    So it is quite straightforward to compute the gradients
+    of the `best score` ``s`` with respect to ``a``, ``b``, ``c`` and ``d``.
+
+    .. math::
+
+      \frac{\partial s}{\partial a} = 0
+
+      \frac{\partial s}{\partial b} = \frac{\partial (b + d)}{\partial b} = 1
+
+      \frac{\partial s}{\partial c} = 0
+
+      \frac{\partial s}{\partial b} = \frac{\partial (b + d)}{\partial d} = 1
```

Comment:
should be ` \frac{\partial s}{\partial d}` here

Reply:
Thanks, fixed.